Rover: improved handling of reverse

rmackay9 opened the pull request:
This PR improves Rover's handling of reversing including fixing this issue (https://github.com/ArduPilot/ardupilot/issues/8311):

- AP_MotorsUGV always interprets a positive steering input as a request to turn the vehicle clockwise regardless of the throttle input.  Essentially this change involves making Ackermann steering vehicles like skid-steering vehicles instead of the other way around.  This simplifies higher level steering controllers because it reduces the number of places we need to check if the vehicle is moving forwards or backwards.
- In manual/semi-manual modes (manual, acro, steering) when interpreting the driver's steering input (which can depend upon the throttle input) we use a throttle-slew or acceleration-limited throttle.  This resolves an issue in which the vehicle might temporarily turn in the wrong direction as the vehicle transitioned between forward and reverse.
- speed scaling of steering for Ackermann steer vehicles is done in the motors library instead of the attitude controller.  This has the following benefits:
   - architecturally puts the scaling very close to the similar scaling for vectored-thrust vehicles
   - allows support for vehicles with both skid-steering and Ackermann style steering (like a ski-doo).
   - essentially scales down the I-term contribution to steering as speed increases.  previously when done in the attitude controller, the scaling was done on the error term.  This led to I-term built up at low speeds taking longer to "bleed off" at high speeds.
- the potential downside of the above change is that at very high speeds the steering control may become so limited that it cannot counteract a physical misalignment in the steering linkage.  For example at 20m/s (70km/h) only 5% of the total steering range is available.  A potential solution to this though is to allow steering inputs > 1.
- a relatively minor change included in this PR is that we perform no steering scaling in manual mode.  this change only affects the relatively new vector-thrust feature that has not yet been released.

OXINARF commented on APMrover2/mode.cpp:
```diff
@@ -90,6 +90,40 @@ void Mode::get_pilot_desired_steering_and_throttle(float &steering_out, float &t
+    float throttle_out_limited = g2.motors.get_slew_limited_throttle(throttle_out, rover.G_Dt);
+    if ((is_negative(throttle_out) != is_negative(throttle_out_limited)) &&
+        ((g.pilot_steer_type == PILOT_STEER_TYPE_DEFAULT) ||
+         (g.pilot_steer_type == PILOT_STEER_TYPE_DEFAULT))) {
```
Something is wrong here, right? The second one was supposed to be PILOT_STEER_TYPE_DIR_REVERSED_WHEN_REVERSING?

OXINARF commented on APMrover2/mode.cpp:
```diff
@@ -90,6 +90,40 @@ void Mode::get_pilot_desired_steering_and_throttle(float &steering_out, float &t
 
+// decode pilot steering held in channel_steer, channel_throttle and return in steer_out and throttle_out arguments
+// steering_out is in the range -4500 ~ +4500 with positive numbers meaning rotate clockwise
+// throttle_out is in the range -100 ~ +100
```
Nitpicking: making the comment from header be the same here would be better.

OXINARF commented on APMrover2/mode.cpp:
```diff
@@ -90,6 +90,40 @@ void Mode::get_pilot_desired_steering_and_throttle(float &steering_out, float &t
+    float speed_out_limited = g2.attitude_control.get_desired_speed_accel_limited(speed_out);
+    if ((is_negative(speed_out) != is_negative(speed_out_limited)) &&
+        ((g.pilot_steer_type == PILOT_STEER_TYPE_DEFAULT) ||
+         (g.pilot_steer_type == PILOT_STEER_TYPE_DEFAULT))) {
```
Same.

rmackay9 replied:
yes, a bug, thanks again.  fixed in a follow up commit

OXINARF commented on APMrover2/mode_steering.cpp:
```diff
@@ -15,38 +15,41 @@ void ModeSteering::update()
     float desired_steering, desired_speed;
     get_pilot_desired_steering_and_speed(desired_steering, desired_speed);
 
+    bool reversed = is_negative(desired_steering);
```
Is this really supposed to be `desired_steering` and not `desired_speed`?

rmackay9 replied:
That was definitely a bug, fixed in follow up commit.

rmackay9 commented:
@OXINARF, thanks for the review!  As per usual you've found a couple of mistakes I've made, having your review really gives me confidence.  Txs and fixes coming shortly.

rmackay9 commented on APMrover2/AP_MotorsUGV.cpp:
```diff
@@ -524,16 +537,12 @@ void AP_MotorsUGV::output_throttle(SRV_Channel::Aux_servo_function_t function, f
+    const float throttle_orig = _throttle;
+    _throttle = get_slew_limited_throttle(_throttle, dt);
+    if (_throttle < throttle_orig) {
+        limit.throttle_upper = true;
```
OK, reversed the conditional command to make it easier to understand, thanks!

rmackay9 commented on APMrover2/AP_MotorsUGV.cpp:
```diff
@@ -151,9 +152,12 @@ void AP_MotorsUGV::setup_servo_output()
 // set steering as a value from -4500 to +4500
-void AP_MotorsUGV::set_steering(float steering)
+//   apply_scaling should be set to false for manual modes where
+//   no scaling by speed or angle should e performed
```
cool, fixed the spelling mistake, txs

rmackay9 commented on APMrover2/AP_MotorsUGV.h:
```diff
@@ -115,4 +119,8 @@ class AP_MotorsUGV {
+
+    // steering scaling related variables
+    bool    _scale_steering = true; // true if we should scale steering by speed or angle
+    float   _ground_speed;  // speed of vehicle over surface.  used to scale back steering at high speed or when reversing
```
yup, removed the ground_speed member and we just pass it in each time.

khancyr commented:
Humm, it looks nice, but I will need a little more time to review correctly. 
One point that bother me is that steering is diverging from the Auto controler when then should be close to allow debugging the auto controler.
I will try to put more test on rover to check the controller in autotest

rmackay9 commented:
I've updated this PR with the following changes:

- rebased on master to resolve conflicts with the omni-rover PR
- fixed a bug in the "Rover: fix calc-steering-to-waypoint to use real heading without reverse" that was causing missions to fail for skid-steering vehicles.

rmackay9 commented:
I tested this some more yesterday on real vehicles and it seems OK so merging!
